### PR TITLE
*: backup to S3 and restore from S3 #116

### LINKED ISSUE #116: [feature]add xtrabackup manger
**Is your feature request related to a problem? Please describe.**

we need xtrabackup to cloud for the cluster
**Describe the solution you'd like**
Make a backup like this:
```shell
kubectl apply -f mysql_v1alpha1_backup.yaml 
```
To backup the database data to S3 storage , or Persisten volume
Then could use the backup copy to create cluster , add RestoreFrom feild in mysql_v1alpha1_cluster.yaml
```shell
kubectl apply -f mysql_v1alpha1_cluster.yaml 
```
create the cluster from backup copy

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by acekingke**
### What type of PR is this?

/enhancement
### Which issue(s) this PR fixes?

Fixes #116 

### What this PR does?

Summary:
use S3 backup cluster and restore cluster from S3

### CODE REVIEW

File: PROJECT [hunk 1]
```diff
@@ -18,4 +18,13 @@ resources:
   group: mysql
   kind: Status
   version: v1alpha1
+- api:
+    crdVersion: v1
+    namespaced: true
+  controller: true
+  domain: radondb.io
```
Comment (andyli029) [inline PROJECT:25]: radondb.io
->
radondb.com

File: api/v1alpha1/backup_types.go [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+	Status BackupStatus `json:"status,omitempty"`
+}
+
+// BackupCondition defines condition struct for backup resource
+type BackupCondition struct {
+	// type of cluster condition, values in (\"Ready\")
+	Type BackupConditionType `json:"type"`
+	// Status of the condition, one of (\"True\", \"False\", \"Unknown\")
+	Status corev1.ConditionStatus `json:"status"`
+
```
Comment (andyli029) [inline api/v1alpha1/backup_types.go:68]: Delete the space

File: config/manager/kustomization.yaml [hunk 3]
```diff
@@ -5,6 +5,12 @@ generatorOptions:
...
-- name: manager-config
-  files:
+- files:
   - controller_manager_config.yaml
+  name: manager-config
+apiVersion: kustomize.config.k8s.io/v1beta1
+kind: Kustomization
+images:
+- name: controller
+  newName: acekingke/controller
```
Comment (andyli029) [inline config/manager/kustomization.yaml:15]: Fix the image.

File: config/manager/kustomization.yaml [hunk 4]
```diff
@@ -7,4 +7,4 @@ generatorOptions:
 configMapGenerator:
 - name: manager-config
   files:
-  - controller_manager_config.yaml
+  - controller_manager_config.yaml
```
Comment (zhyass) [inline config/manager/kustomization.yaml:10]: format

File: cmd/sidecar/main.go [hunk 5]
```diff
@@ -51,10 +52,61 @@ func main() {
...
+		Use:   "backup",
+		Short: "Take a backup from node and push it to rclone path.",
+		Args: func(cmd *cobra.Command, args []string) error {
+			if len(args) != 1 {
+				return fmt.Errorf("require one arguments. source host and destination bucket")
+			}
+			return nil
+		},
+		Run: func(cmd *cobra.Command, args []string) {
+			//TODO: make backup cmd
```
Comment (zhyass) [inline cmd/sidecar/main.go:68]: remove the unused code

File: cmd/sidecar/main.go [hunk 6]
```diff
@@ -51,10 +52,61 @@ func main() {
...
+			}
+		},
+	}
+	cmd.AddCommand(takeBackupCmd)
+	httpCmd := &cobra.Command{
+		Use:   "http",
+		Short: "start http server",
+
+		Run: func(cmd *cobra.Command, args []string) {
+			//TODO: make backup cmd
```
Comment (zhyass) [inline cmd/sidecar/main.go:83]: remove the unused code

File: cmd/sidecar/main.go [hunk 7]
```diff
@@ -51,10 +52,61 @@ func main() {
...
+		Use:   "request_a_backup",
+		Short: "start request a backup",
+		Args: func(cmd *cobra.Command, args []string) error {
+			if len(args) != 1 {
+				return fmt.Errorf("require one arguments. ")
+			}
+			return nil
+		},
+		Run: func(cmd *cobra.Command, args []string) {
+			//TODO request a backup
```
Comment (zhyass) [inline cmd/sidecar/main.go:102]: remove the unused code

File: backup/syncer/job.go [hunk 8]
```diff
@@ -0,0 +1,142 @@
...
+limitations under the License.
+*/
+
+package syncer
+
+import (
+	"fmt"
+
+	"github.com/presslabs/controller-util/syncer"
+	batch "k8s.io/api/batch/v1"
```
Comment (zhyass) [inline backup/syncer/job.go:23]: maybe `batchv1` is much better?

File: backup/status.go [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+limitations under the License.
+*/
+
+package backup
+
+import (
+	"fmt"
+	"time"
+
+	core "k8s.io/api/core/v1"
```
Comment (zhyass) [inline backup/status.go:23]: maybe use `corev1` is much better

File: backup/status.go [hunk 10]
```diff
@@ -0,0 +1,88 @@
...
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	apiv1 "github.com/radondb/radondb-mysql-kubernetes/api/v1alpha1"
```
Comment (zhyass) [inline backup/status.go:26]: `apiv1alpha1`

File: api/v1alpha1/backup_types.go [hunk 11]
```diff
@@ -0,0 +1,97 @@
...
+type BackupSpec struct {
+	// INSERT ADDITIONAL SPEC FIELDS - desired state of cluster
+	// Important: Run "make" to regenerate code after modifying this file
+	// To specify the image that will be used for sidecar container.
+	// +optional
+	// +kubebuilder:default:="acekingke/sidecar:0.1"
+	Image string `json:"image"`
+	// HostName represents the host for which to take backup
+	HostName    string `json:"hostname"`
+	ClusterName string `json:"clustname"`
```
Comment (zhyass) [inline api/v1alpha1/backup_types.go:37]: Add comment for `ClusterName`

**Review by zhyass — COMMENTED** [review]
Great job !

**Comment by andyli029** [comment]
Fix the places pointed out by others first:
https://github.com/radondb/radondb-mysql-kubernetes/pull/138

File: api/v1alpha1/backup_types.go [hunk 12]
```diff
@@ -0,0 +1,98 @@
...
+)
+
+// EDIT THIS FILE!  THIS IS SCAFFOLDING FOR YOU TO OWN!
+// NOTE: json tags are required.  Any new fields you add must have json tags for the fields to be serialized.
+
+// BackupSpec defines the desired state of Backup
+type BackupSpec struct {
+	// INSERT ADDITIONAL SPEC FIELDS - desired state of cluster
+	// Important: Run "make" to regenerate code after modifying this file
+	// To specify the image that will be used for sidecar container.
```
Comment (zhyass) [inline api/v1alpha1/backup_types.go:31]: delete `// INSERT ADDITIONAL SPEC FIELDS - desired state of cluster
	// Important: Run "make" to regenerate code after modifying this file` or Skip a line.

File: api/v1alpha1/backup_types.go [hunk 13]
```diff
@@ -0,0 +1,98 @@
...
+	HostName string `json:"hostname"`
+	// Cluster represents the cluster name to backup
+	ClusterName string `json:"clustname"`
+}
+
+// BackupStatus defines the observed state of Backup
+type BackupStatus struct {
+	// INSERT ADDITIONAL STATUS FIELD - define observed state of cluster
+	// Important: Run "make" to regenerate code after modifying this file
+	Completed bool `json:"completed,omitempty"`
```
Comment (zhyass) [inline api/v1alpha1/backup_types.go:45]: Skip a line and add comment

File: backup/syncer/job.go [hunk 14]
```diff
@@ -0,0 +1,140 @@
...
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+
+	v1alpha1 "github.com/radondb/radondb-mysql-kubernetes/api/v1alpha1"
+	"github.com/radondb/radondb-mysql-kubernetes/backup"
+)
+
+var log = logf.Log.WithName("backup.syncer.job")
+
+//TODO: sync job
```
Comment (zhyass) [inline backup/syncer/job.go:36]: remove the unused code

File: config/rbac/backup_viewer_role.yaml [hunk 15]
```diff
@@ -0,0 +1,20 @@
+# permissions for end users to view backups.
+apiVersion: rbac.authorization.k8s.io/v1
+kind: ClusterRole
+metadata:
+  name: backup-viewer-role
+rules:
+- apiGroups:
+  - mysql.radondb.io
```
Comment (zhyass) [inline config/rbac/backup_viewer_role.yaml:8]: mysql.radondb.com

File: sidecar/server.go [hunk 16]
```diff
@@ -0,0 +1,176 @@
...
+func checkBackupTrailers(resp *http.Response) error {
+	if values, ok := resp.Trailer[backupStatusTrailer]; !ok || !stringInSlice(backupSuccessful, values) {
+		// backup is failed, remove from remote
+		return fmt.Errorf("backup failed to be taken: no 'Success' trailer found")
+	}
+
+	return nil
+}
+
+func stringInSlice(a string, list []string) bool {
```
[the remark this answers is not in the document]
Comment (zhyass) [reply sidecar/server.go:169]: remove the unused function `stringInSlice`

File: controllers/backup_controller.go [hunk 17]
```diff
@@ -0,0 +1,127 @@
...
+		log.Info("update backup object status")
+		if err := r.Status().Update(context.TODO(), backup.Unwrap()); err != nil {
+			log.Error(err, fmt.Sprintf("update status backup %s/%s", backup.Name, backup.Namespace),
+				"backupStatus", backup.Status)
+			return err
+		}
+	}
+	return nil
+}
+func (r *BackupReconciler) sync(ctx context.Context, syncers []syncer.Interface) error {
```
[the remark this answers is not in the document]
Comment (zhyass) [reply controllers/backup_controller.go:116]: delete the unused function.

File: controllers/backup_controller.go [hunk 18]
```diff
@@ -0,0 +1,127 @@
...
+	for _, s := range syncers {
```
Comment (zhyass) [inline controllers/backup_controller.go:117]: Remove the unused function

File: docs/DEPLOY.md [hunk 19]
```diff
@@ -0,0 +1,115 @@
...
+spec:
+  replicas: 3
+  mysqlVersion: "5.7"
+  backupSecretName: sample-backup-secret
+  ...
+```
+now create backup yaml file mysql_v1_backup.yaml like this:
+
+```yaml
+apiVersion: mysql.radondb.io/v1
```
Comment (zhyass) [inline docs/DEPLOY.md:81]: radondb.com

**Comment by zhyass** [comment]
LGTM

File: docs/DEPLOY.md [hunk 20]
```diff
@@ -0,0 +1,115 @@
+# mysql-operator
```
Comment (andyli029) [inline docs/DEPLOY.md:1]: Add CMD to restore from s3

File: sidecar/config.go [hunk 21]
```diff
@@ -108,9 +149,62 @@ func NewConfig() *Config {
...
+	}
+
+	tmpdir := "/root/backup/"
+	if len(cfg.XtrabackupTargetDir) != 0 {
+		tmpdir = cfg.XtrabackupTargetDir
+	}
+	xtrabackupArgs := []string{
+		"--backup",
+		"--stream=xbstream",
+		"--host=127.0.0.1",
```
Comment (zhyass) [inline sidecar/config.go:184]: Lack password?

File: sidecar/config.go [hunk 22]
```diff
@@ -108,9 +149,62 @@ func NewConfig() *Config {
 
 		AdmitDefeatHearbeatCount: int32(admitDefeatHearbeatCount),
 		ElectionTimeout:          int32(electionTimeout),
+
+		ClusterName:                getEnvValue("SERVICE_NAME"),
+		BackupUser:                 "sys_backups", //getEnvValue("BACKUP_USER"),
```
Comment (zhyass) [inline sidecar/config.go:154]: `BackupUser` and `BackupPassword` cannot directly specify, maybe you can use OperatorUser and OperatorPassword as the backup user and password

File: controllers/backup_controller.go [hunk 23]
```diff
@@ -0,0 +1,114 @@
...
+	if err = r.updateBackup(savedBackup, backup); err != nil {
+		return reconcile.Result{}, err
+	}
+
+	return ctrl.Result{}, nil
+}
+
+// SetupWithManager sets up the controller with the Manager.
+func (r *BackupReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewControllerManagedBy(mgr).
```
Comment (zhyass) [inline controllers/backup_controller.go:92]: return ctrl.NewControllerManagedBy(mgr).
		For(&apiv1alpha1.Backup{}).
                Owns(&batchv1.Job{}).
		Complete(r)

File: docs/DEPLOY.md [hunk 24]
```diff
@@ -0,0 +1,133 @@
...
+
+## Quickstart
```
Comment (runkecheng) [inline docs/DEPLOY.md:3]: cant deploy by this way now, maybe need to update helm

**Comment by andyli029** [comment]
Version: 0.1.88

**Comment by TCeason** [comment]
> Version: 0.1.88

It's really a good pr. When can we use it on master？

**Comment by andyli029** [comment]
> > Version: 0.1.88
> 
> It's really a good pr. When can we use it on master？

Soon.

File: cluster/syncer/statefulset.go [hunk 25]
```diff
@@ -567,3 +575,26 @@ func xenonHttpRequest(host, method, url string, rootPasswd []byte, body io.Reade
 
 	return resp.Body, nil
 }
+
+//check the backup is exist and running
+func (s *StatefulSetSyncer) backupIsRunning(ctx context.Context) (bool, error) {
+
```
Comment (andyli029) [inline cluster/syncer/statefulset.go:581]: Delete the line.

**Comment by TCeason** [comment]
> > > Version: 0.1.88
> > 
> > 
> > It's really a good pr. When can we use it on master？
> 
> Soon.

Pretty good.

File: docs/DEPLOY.md [hunk 26]
```diff
@@ -0,0 +1,133 @@
...
+kubectl create -f config/samples/backup_secret.yaml
+```
+
+```shell
+kubectl apply -f config/samples/mysql_v1alpha1_cluster.yaml     
+```
+## backup
+After run cluster success
+```shell
+kubectl apply -f config/samples/samples/mysql_v1alpha1_backup.yaml
```
Comment (andyli029) [inline docs/DEPLOY.md:22]: config/samples/samples/mysql_v1alpha1_backup.yaml
->
config/samples/mysql_v1alpha1_backup.yaml

File: docs/DEPLOY.md [hunk 27]
```diff
@@ -0,0 +1,126 @@
...
+```shell
+kubectl delete clusters.mysql.radondb.com sample
+```
+
+Uninstall the operator name `test`:
+
+```shell
+helm uninstall test
+```
+
```
Comment (andyli029) [inline docs/DEPLOY.md:86]: Add 
```
kubectl delete -f config/samples/mysql_v1alpha1_backup.yaml
```

**Pull request closed without merging**